src/Chainweb/Pact/PactService.hs

```diff
@@ -1077,26 +1077,29 @@ transactionsFromPayload plData = do
 
 execLookupPactTxs
     :: PayloadCas cas
-    => Maybe (T2 BlockHeight BlockHash)
+    => Rewind (T2 BlockHeight BlockHash)
     -> Vector P.PactHash
     -> PactServiceM cas (Vector (Maybe (T2 BlockHeight BlockHash)))
 execLookupPactTxs restorePoint txs
     | V.null txs = return mempty
     | otherwise = go
   where
-    getRestorePoint = case restorePoint of
-        Nothing -> fmap (\bh -> T2 (_blockHeight bh) (_blockHash bh))
-            <$> findLatestValidBlock
-        x -> return x
+    doRewind cp (T2 lh lha) =
+      withCheckpointerRewind (Just (lh + 1, lha)) "lookupPactTxs" $ \_ ->
+        liftIO $ Discard <$> V.mapM (_cpLookupProcessedTx cp) txs
 
     go = do
-        cp <- getCheckpointer
-        mrp <- getRestorePoint
-        case mrp of
-            Nothing -> return mempty      -- can't look up anything at genesis
-            Just (T2 lh lha) ->
-                withCheckpointerRewind (Just (lh + 1, lha)) "lookupPactTxs" $ \_ ->
-                    liftIO $ Discard <$> V.mapM (_cpLookupProcessedTx cp) txs
+      cp <- getCheckpointer
+      case restorePoint of
+        NoRewind _ -> liftIO $!
+          V.mapM (_cpLookupProcessedTx cp) txs
+        DoRewind a -> doRewind cp a
+        _ -> do
+          mrp <- fmap (\(!bh) -> T2 (_blockHeight bh) (_blockHash bh))
+            <$!> findLatestValidBlock
+          case mrp of
+            Nothing -> return mempty -- nothing denotes genesis
+            Just a -> doRewind cp a
```

Comment:
If `NoRewind`, lookup all txs without rewinding. This is only used when doing SPV lookups. When `DoRewind` or `Genesis`, do the rewind.

Reply:
PreInsert shouldn't rewind, that's a regression -- everything that was `Left` before should be `NoRewind`.